core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ImplementTableFunctionSource.java

```diff
@@ -0,0 +1,500 @@
+        // of each input table come before the "filler" rows.
+        Optional<OrderingScheme> finalOrderBy = Optional.of(new OrderingScheme(ImmutableList.of(finalRowNumberSymbol), ImmutableMap.of(finalRowNumberSymbol, ASC_NULLS_LAST)));
+
+        // Partition the combined sources by their combined partitioning lists.
+        List<Symbol> partitionBy = node.getTableArgumentProperties().stream()
+                .map(TableArgumentProperties::getSpecification)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .map(DataOrganizationSpecification::getPartitionBy)
+                .flatMap(Collection::stream)
+                .collect(toImmutableList());
+
+        PlanNode pruned = pruneHelperSymbols(marked.getNode(), partitionSizeSymbols.build(), rowNumberSymbols.values(), combinedRowNumberSymbols.build(), finalRowNumberSymbol, context);
+
+        // Combine the pass through specifications from all sources
+        List<PassThroughSpecification> passThroughSpecifications = node.getTableArgumentProperties().stream()
+                .map(TableArgumentProperties::getPassThroughSpecification)
+                .collect(toImmutableList());
+
+        // Combine the required symbols from all sources
+        List<List<Symbol>> requiredSymbols = node.getTableArgumentProperties().stream()
+                .map(TableArgumentProperties::getRequiredColumns)
+                .collect(toImmutableList());
+
+        return Result.ofPlanNode(new TableFunctionProcessingNode(
+                node.getId(),
+                node.getName(),
+                node.getProperOutputs(),
+                Optional.of(pruned),
+                passThroughSpecifications,
+                requiredSymbols,
+                Optional.of(markerSymbols),
+                Optional.of(new DataOrganizationSpecification(partitionBy, finalOrderBy)),
+                ImmutableSet.of(),
+                0,
+                Optional.empty(),
+                node.getHandle()));
+    }
+
+    private static NodeWithSymbols appendHelperSymbolsForSource(
+            int index,
+            TableFunctionNode node,
+            ImmutableList.Builder<Symbol> partitionSizeSymbols,
+            ImmutableMap.Builder<Symbol, Symbol> rowNumberSymbols,
+            Metadata metadata,
+            Context context)
+    {
+        String argumentName = node.getTableArgumentProperties().get(index).getArgumentName();
+        PlanNode source = node.getSources().get(index);
+
+        Symbol rowNumber = context.getSymbolAllocator().newSymbol(argumentName + "_row_number", BIGINT);
+        source.getOutputSymbols().stream()
+                .forEach(symbol -> rowNumberSymbols.put(symbol, rowNumber));
+        WindowNode.Function rowNumberFunction = new WindowNode.Function(
+                metadata.resolveFunction(context.getSession(), QualifiedName.of("row_number"), ImmutableList.of()),
```

Comment:
Resolving functions is generally not very cheap, so it might make sense to do it once and reuse.